Archive install fails on Debian 9 and Ubuntu.  Have added new variable for SystemD path.

JumaX opened the pull request:
# Description

When using the Archive installer on Debian 9, it fails as the following path does not exist by default on this Linux distribution:

```fatal: [zookeeper1]: FAILED! => {"changed": false, "msg": "Destination directory /usr/lib/systemd/system does not exist"}```

I've added a task that specifically triggers when an archive install is run and its against Debian, whereby the above path is created.

Fixes # (issue)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

I have added a new scenario called archive-plain-debian which tests the installation on Debian9 with TLS and SASL PLAIN enabled.

**Test Configuration**:

The test spins up a complete cluster using archive installation with sasl plain and TLS enabled.  There is a verfication that properties for SASL_SSL have been configured across all components, in addition to health checks that validate the services are all running.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] Any variable changes have been validated to be backwards compatible

erikgb commented on roles/confluent.control_center/tasks/main.yml:
```diff
@@ -61,6 +61,15 @@
     recurse: true
 
+# Archive File deployments need to create SystemD directory on Debian
+- name: Create Zookeeper Service path
```
I would suggest to put this repetitive task in common-role. Don't we already have OS-specific tasks there?

JumaX replied:
Really good idea!  Will change shortly.

erikgb commented on roles/confluent.control_center/tasks/main.yml:
```diff
@@ -61,6 +61,15 @@
 
+# Archive File deployments need to create SystemD directory on Debian
+- name: Create Zookeeper Service path
+  ansible.builtin.file:
```
```suggestion
  file:
```
We should migrate to FQN, but for now I would prefer consistency.

JumaX replied:
Yeah, the latest docs actually state to use FQDN: https://docs.ansible.com/ansible/latest/collections/ansible/builtin/file_module.html

There are two approaches here, either we kick off a full project to change this across the entire project, or we start changing it as we go along.

To me, it would make sense to change it as we go along, as it would be less work for anyone person.

erikgb replied:
That could also be a working path, and I am happy with either. What does @domenicbove thinks?

JumaX replied:
@erikgb I can confirm that right now we don't have the resources internally to kick off a dedicated project for this, however if you wanted to go through and update them all I wouldn't be opposed. :).

erikgb replied:
My personal (and company) priority now is collection compatible layout (https://github.com/confluentinc/cp-ansible/pull/630) - to avoid the pesky submodules. And that will require a massive change in a lot of files.

As I said, I would be happy with introducing FQN gradually. If so, I think it should be mentioned in the development guidelines. And consider non-FQN as technical debt.

JumaX replied:
@erikgb I've changed it back to just file: and have added an internal task for us to review changing this across the board in the future.  Thanks again for the discussion, it makes sense to keep it consistent for the wider community for now.

erikgb replied:
IMO GitHub issues should be created for issues/tasks that are relevant for contributors and the community in general. I understand that Confluent also must track some internal work - based on tickets from supported customers. But I think it is more "community friendly" to consider GitHub issues as the master for issues/tasks in this project.

JumaX replied:
@erikgb good feedback.  It's something we've discussed internally before, but still not sure we have the correct direction here.  It's a constant challenge between what we keep internal, vs what is public.  For example, we have some proprietary features, that we may not want to have any public details on immediately, so we need to find a way to manage that properly.  I'll let our Product Management team know about your thoughts.

erikgb commented on roles/confluent.control_center/tasks/main.yml:
```diff
@@ -61,6 +61,15 @@
+# Archive File deployments need to create SystemD directory on Debian
+- name: Create Zookeeper Service path
+  ansible.builtin.file:
+    path: "/usr/lib/systemd/system"
```
I think it is considered best practice to set owner, group and permissions.

JumaX replied:
It adopts the OS defaults, and because this is a system directory it defaults to the following which is what is desired: 

drwxr-xr-x 2 root root 4096 Apr 20 09:26 system

So we can specify the owner group and permissions, but it's not required in this instance.

domenicbove commented:
Hey, so @nerdynick fixed this bug in 6.2.x by creating the `systemd_base_dir` variable:

```
systemd_base_dir: "{{'/lib/systemd/system' if ansible_os_family == 'Debian' else '/usr/lib/systemd/system'}}"
zookeeper:
  systemd_file: "{{systemd_base_dir}}/{{zookeeper_service_name}}.service"
```

https://github.com/confluentinc/cp-ansible/blob/6.2.x/roles/confluent.variables/vars/main.yml#L4

I think we should try to use her same solution in 6.1.x to avoid merge conflicts when merging up.

erikgb reviewed (commented):
It seems like this PR just adds test code/configuration? If that is correct, I think the PR title and description should be updated.

And personally I think a lot of test/code configuration is added. Is there a way to compact the change. I think it is even hard to review. 😅

JumaX commented:
@erikgb Not completely correct no. This PR added additional variables and updated the path to use a variable for systemD configuration as suggested by @domenicbove previously.  However it looks like the following PR was merged into 6.1.x yesterday, which also added this: https://github.com/confluentinc/cp-ansible/pull/647

We have a practice of merging the latest 6.1.x changes into our PRs, to resolve any merge conflicts, before then merging back into the 6.1.x branch.  Because of this, the changes to the variables and paths for systemD do not show up.

I hope that makes sense, it took me a bunch of time this morning to figure out what happened to my changes. :)

domenicbove commented:
Oh wow, @JumaX I made a mistake and should not have merge this PR: https://github.com/confluentinc/cp-ansible/pull/647

I thought it was only a change to the dev guide, not a big change! Please bare with my while I correct this

domenicbove commented:
@JumaX Really sorry for the mistake. This is totally my fault. I should not have merge that PR. Can you undo the merge of 6.1.x into your branch? Fortunately, it should be revertable with:
```
git log
git reset --hard <previous commit>
git push
```

JumaX commented:
@domenicbove all good, and back to normal.

erikgb reviewed (approved):
LGTM; now I can see what I am supposed to review. 😉 The current change answers a question I had when browsing through the previous suggestion. It seems like Debian hasn't finalized the /usr/lib merge, and now this is clear from the code changes.

I still think the amount of test-code added in this PR is big, but I am not yet familiar with the molecule setup in this project. I have never been able to run the tests locally.... 😢

JumaX commented:
@erikgb agreed its a lot for the test code.  The upside, it's pretty much boiler plate, so actually doesn't need to be reviewed that heavily, its the non test code we are mostly concerned about.
